File: scala/src/test/scala/com/example/cache/CacheUtilsTest.scala


package com.example.cache

import java.util.concurrent.atomic.AtomicBoolean

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

@Test
class CacheUtilsTest {

    @Test
    def emptyCacheGet() = {
        val cache = scala.collection.mutable.Map.empty[Int, Int]
        assertTrue(CacheUtils.get(cache).apply(1).isEmpty)
    }

    @Test
    def fullCacheGet() = {
        val cache = scala.collection.mutable.Map.empty[Int, Int]
        cache.put(1, 1)
        assertFalse(CacheUtils.get(cache).apply(1).isEmpty)
    }
    
    @Test
    def cacheMemoize() = {
        val cache = scala.collection.mutable.Map.empty[Int, Int]
        val get = CacheUtils.get(cache);
        val put = CacheUtils.put(cache);
        
        val hit = new AtomicBoolean();
        val f = (a: Int) => {
            hit.set(true)
            a
        }
        val g = CacheUtils.memoize(get, put, f)
        
        hit.set(false)
        assertTrue(f.apply(1) == 1)
        assertTrue(hit.get);

        hit.set(false)
        assertTrue(g.apply(1) == 1)
        assertTrue(hit.get);
        
        hit.set(false)
        assertTrue(g.apply(1) == 1)
        assertFalse(hit.get);

        hit.set(false)
        assertTrue(g.apply(1) == 1)
        assertTrue(hit.get);
    }
}

File: java/src/test/java/com/example/cache/CacheUtilsTest.java
package com.example.cache;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.BiConsumer;
import java.util.function.Function;

import org.junit.Test;

public class CacheUtilsTest {

    @Test
    public void emptyCacheGet() {
        Map<Integer, Integer> cache = new HashMap<>();
        assertFalse(CacheUtils.get(cache).apply(1).isPresent());
    }

    @Test
    public void fullCacheGet() {
        Map<Integer, Integer> cache = new HashMap<>();
        cache.put(1, 1);
        assertTrue(CacheUtils.get(cache).apply(1).isPresent());
    }
    
    @Test
    public void cacheMemoize() {
        Map<Integer, Integer> cache = new HashMap<>();
        Function<Integer, Optional<Integer>> get = CacheUtils.get(cache);
        BiConsumer<Integer, Optional<Integer>> put = CacheUtils.put(cache);
        
        AtomicBoolean hit = new AtomicBoolean();
        Function<Integer, Integer> f = (Integer a) -> {
            hit.set(true);
            return a;
        };
        Function<Integer, Integer> g = CacheUtils.memoize(get, put, f);
        
        hit.set(false);
        assertTrue(f.apply(1).equals(1));
        assertTrue(hit.get());
        
        hit.set(false);
        assertTrue(g.apply(1).equals(1));
        assertTrue(hit.get());

        hit.set(false);
        assertTrue(g.apply(1).equals(1));
        assertFalse(hit.get());
        
        hit.set(false);
        assertTrue(g.apply(2).equals(2));
        assertTrue(hit.get());
    }
}
